src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTag.cs

```diff
@@ -0,0 +1,199 @@
+        /// </summary>
+        public override GraphicsResult GetGraphics(IWpfTextView view, Geometry bounds, TextFormattingRunProperties format)
+        {
+            var block = new TextBlock
+            {
+                FontFamily = format.Typeface.FontFamily,
+                FontSize = 0.75 * format.FontRenderingEmSize,
+                FontStyle = FontStyles.Normal,
+                Foreground = format.ForegroundBrush,
+                Padding = new Thickness(left: 2, top: 0, right: 2, bottom: 0),
+            };
+
+            var idRun = GetRunForId(out var hyperlink);
+
+            if (hyperlink is null)
+            {
+                block.Inlines.Add(idRun);
+            }
+            else
+            {
+                block.Inlines.Add(hyperlink);
+                hyperlink.RequestNavigate += HandleRequestNavigate;
+            }
+
+            block.Inlines.Add(": " + _diagnostic.Message);
+
+            var lineHeight = Math.Floor(format.Typeface.FontFamily.LineSpacing * block.FontSize);
+            var image = new CrispImage
+            {
+                Moniker = GetMoniker(),
+                MaxHeight = lineHeight,
+                Margin = new Thickness(1, 0, 5, 0)
+            };
+
+            var border = new Border
+            {
+                BorderBrush = format.BackgroundBrush,
+                BorderThickness = new Thickness(1),
+                Background = Brushes.Transparent,
+                Child = new StackPanel
+                {
+                    Height = lineHeight,
+                    Orientation = Orientation.Horizontal,
+                    Children = { image, block }
+                },
+                CornerRadius = new CornerRadius(2),
+                // Highlighting lines are 2px buffer. So shift us up by one from the bottom so we feel centered between them.
+                Margin = new Thickness(10, top: 0, right: 0, bottom: 1),
+                Padding = new Thickness(1)
+            };
+
+            // This is used as a workaround to the moniker issues in blue theme
+            var editorBackground = (Color)_editorFormatMap.GetProperties("TextView Background")["BackgroundColor"];
+            ImageThemingUtilities.SetImageBackgroundColor(border, editorBackground);
+
+            border.Measure(new Size(double.PositiveInfinity, double.PositiveInfinity));
+            view.ViewportWidthChanged += ViewportWidthChangedHandler;
+            view.LayoutChanged += View_LayoutChanged;
+
+            return new GraphicsResult(border, dispose:
+                () =>
+                {
+                    if (hyperlink is not null)
+                    {
+                        hyperlink.RequestNavigate -= HandleRequestNavigate;
+                    }
+
+                    view.ViewportWidthChanged -= ViewportWidthChangedHandler;
+                    view.LayoutChanged -= View_LayoutChanged;
+                });
+
+            void ViewportWidthChangedHandler(object s, EventArgs e)
```

Comment:
could you doc why both the tag *and* the adornmetn manager listen for hte viewport width cahnging?